### PR TITLE
build JupyterHub singleuser images for each notebook stack

### PR DESCRIPTION
**Pull request by minrk**
For every `*-notebook` stack, a `*-singleuser` image is also built

cf jupyter/dockerspawner#82

### CODE REVIEW

**Comment by parente** [comment]
Nice. Will wait to see how it fares on travis.

Note to self / maintainers: make sure we manually create the repos on Docker Hub and assign the proper build permissions before merging and running `make release`. (The root README here has the details in the "new stack" section).

**Comment by parente** [comment]
We should also mention these both in the top-level README in the tip section and within each *-notebook README. What they're for, how to use them with JupyterHub, pointer to JupyterHub doc about singleuser.

**Comment by minrk** [comment]
Looks like I've missed something in the Makefile

**Comment by parente** [comment]
> docker build  --rm --force-rm -t jupyter/minimal-singleuser:latest ./minimal-singleuser

It's trying to build in a folder named `./minimal-singleuser` but it looks like the bash script reuses the folder `singleuser=dockerspawner/singleuser`

**Comment by jakirkham** [comment]
@parente, I don't believe I have permissions on Docker Hub. Is there some email I missed where I needed to accept something?

**Comment by parente** [comment]
@jakirkham You're all set now. Part of the "stacks" team.

**Comment by jakirkham** [comment]
Just got the notification. Thanks @parente.

**Comment by rgbkrk** [comment]
Looks like binstar went out during the last build. Is this good to go otherwise?

**Comment by parente** [comment]
I think the DH projects still need to be created before we can push.

**Comment by minrk** [comment]
I *think* so, but let's give Travis one more shot. I managed to introduce a BSD-sed vs GNU-sed incompatibility in one of the last builds.

@parente I made the DH projects just now, I think.

File: Makefile [hunk 1]
```diff
@@ -12,6 +12,11 @@ ALL_STACKS:=minimal-kernel \
 	datascience-notebook \
 	pyspark-notebook \
 	all-spark-notebook
+
+ALL_SINGLEUSERS:=$(shell echo $(ALL_STACKS) | sed "s/ /\n/g" | grep notebook | sed s/notebook/singleuser/g)
```
Comment (parente) [inline Makefile]: echoing this gives me `minimal-kernelnminimal-singleusernr-singleusernscipy-singleuserndatascience-singleusernpyspark-singleusernall-spark-singleuser`.

Lack of whitespace is going to screw up later steps?

Comment (minrk) [inline Makefile]: There isn't a lack of whitespace, they are newlines instead of spaces.

**Comment by parente** [comment]
@minrk See them now. Can you copy/paste the long description from the exist repos and adjust accordingly? Don't want to recreate the docker-hell of unknown images we had before. (The README in this repo has the steps in the Maintainer section.)

One more comment about the Makefile.

IMHO, I think we should wait til Travis makes it all the way through before merging this. There's quite a bit of magic in the makefile now. Either we debug it when travis fails or we debug it after merge and when the manual build fails later.

**Comment by minrk** [comment]
DockerHub descriptions updated.

**Comment by parente** [comment]
Travis is green. Output in the logs looks right to me. @minrk can you confirm?

**Comment by minrk** [comment]
Yes, I think so.

**Comment by parente** [comment]
Running the build /  push now.

**Comment by jakirkham** [comment]
I might do another release after you. :wink: We have some changes for datascience that we would like to ship. ( https://github.com/jupyter/docker-stacks/pull/166 )

**Comment by parente** [comment]
@jakirkham That's fine. We should be doing a separate build for each PR anyway. That way if something goes wrong, we know what change broke things.

**Comment by parente** [comment]
Permission problems. `stacks` doesn't have access on the new repos.  Fixing now.

**Comment by parente** [comment]
It's pushing now. Apparently, even though the filesystem layers already exist in other jupyter/* repositories on DH, they still have to push again to the new repos.

Disk is free, right? :money_with_wings:

**Comment by jakirkham** [comment]
Went back and checked a few times. Seemed to have some issues pushing, but we did get to deploy in the last one. So, it should all be up there.

**Pull request merged**